### ISSUE TITLE
Use for small molecules and Volume reconstruction

### ISSUE THREAD

**Issue by DAlvGar** [biocryst/biozernike#21]
Hi! Thanks for the awesome library!  Crunching all those numbers and functions from the paper and putting this together is amazing.  

I am trying to adjust the library to use it for small molecule 3D interaction potential comparison instead of protein densities for my research. I am coming across several challenges to understand all the code and the science behind and I would like to kindly ask for some guidance :)

These interaction potentials contain energy values ranging from negative to positive, although I would not mind to throw away the positive part and convert the negative to positive to mimic the current implementation for densities. These potentials come as DX format grid files.  I already have a parser in java for them.

What are the main things I should take care of in the code ? I saw the similarity function was somehow “trained” but could not find any details. 

I was wondering if you have any example of script on how to reconstruct the volume from the moments as you showcase in figure 1 of your paper for protein 4HHB?

**Comment by biocryst** [comment]
Hi Daniel,

Thank you for your interest!

Sorry for late response, these are busy days for me. To address some of your questions I committed (very dirty) experimental code that handles reconstruction (of ligands no less).  Please check out branch reconstruct_tmp, and an example in VolumeTest.testReconstruction()

You will need to save the cache file once (takes some time), and then it can be used to speed things up. The cache location is hardcoded in the ReconstructionCache class. The file will load rather slowly still (but only once per run), I had issues with faster serialization libraries so left it as is for now. 

In theory Zernike moments will approximate any function, positive and negative values. But in this code we're kinda assuming only positive values are valuable, so the volume normalisation will break. 

Pay attention to the Volume class: the center, radius of gyration and volume mass are the things that will definitely be affected. The first two need to be calculated on the absolute values of the density, for the mass I'm not sure - try converting/reconstructing and see what you get.

The later normalisation (as in, Zernike moment alignment by zero'ing some moment values) should work as usual.

I hope this helps. 

Happy new year!

Kind regards,
Dmytro.

**Comment by DAlvGar** [comment]
Wow!  This really helped a lot! I could expand and reconstruct the energy grids I am dealing with and they make sense. I could also superpose some molecules using the invariants!  I am very grateful! 

I still need to figure out how to get some good reconstructued values which are in the same scale as the input, but I think for my immediate purpose I could get this working quite well! Amazing!

Thanks a lot Dmytro.

**Comment by DAlvGar** [comment]
Hi again, 

it's me now trying to understand the rotation part of the code.... I have problems undertanding how the superposition works... could i get some help understanding this chunk of code below from the `VolumeTest.testVectors` test, for instance? Where do these parameters 2, 2, and indsPositive come from and what do they mean?

Once superposed, i extract the rotated moments somehow? I would like to compare distances taking all the coefficients (and not just the invariants) once rotated. Is that possible? 

Thanks for the invaluable help.
Regards,

Daniel

[(https://github.com/biocryst/biozernike/blob/6dfb2f9fefc0e87909c0ebbbc1b17e7b391fba71/src/test/java/org/rcsb/biozernike/VolumeTest.java#L353-L360)

**Comment by DAlvGar** [comment]
This is super useful! It helped me understanding so many things on the paper and the code!
I appreciate your time, it's very kind of you. 

I have one "maybe last :)" quesntion on the search with the descriptors. I would like to implement something similar where i would have a database with ligands and descriptors to search against. Is the descriptor a plain concatenation of all the CN 2, 3, 4 and 5th invariants (+de GEO part)? And then the weighted distance metric you parameterized has weights for each coefficient?

**Comment by biocryst** [comment]
Indeed. You calculate the moments up to a certain order (it doesn't need to be 20, we just used that number to compare to state of the art), concatenate averaged normalisations for a few orders (InvariantNorm.getInvariants() is doing that) and the geometric descriptor (Descriptor.calcGeometryDescriptor()).

You may need to simplify the geometry descriptor for small molecules. I mean, for example, kurtosis of a distance distribution with 10 points wouldn't make much sense. Maybe the radius and the nominal weight is all you need.

The geometric parameters may be used as a filter (e.g., if any of them differ by more than 20%, then it's not a hit) and then compare moments of the structures that has passed the filter. At the first iteration of your implementation you don't really need to train weights for comparing moments. A regular distance like braycurtis or canberra usually works well enough.

The weights were needed because 1) the geometric descriptor and the moment descriptor are on different scales, and we wanted one distance measure that uses them all, and 2) some moments are more important than others (e.g. the same difference in order 2 and order 20 means vastly different things). It also helps to reduce the final vector quite a bit (if you constrain the weights to >=0, many of them will be 0) - but that will depend on the dataset and the training method.
